feat(core): Add retry execution endpoint to public api

konstantintieber opened the pull request:
## Summary

Add retry execution endpoint to public API under `/api/v1/executions/:id/retry`. Such an endpoint has been only available internally until now.
Adds new permission scope `execution:retry`.

This PR is waiting for the sub-PR with a refactoring of the `ExecutionService.retry` method to be merged into it: https://github.com/n8n-io/n8n/pull/19293
Update Sep 9th: it has been merged into this one.

## 🎥 Demo video

Watch on Loom: https://www.loom.com/share/9a767e11a6ec4aa7b6ce1ec33ffe5fae?sid=fe416b75-8382-4251-9846-7cec94ab89c2




## Related Linear tickets, Github issues, and Community forum posts

Closes: PAY-3649

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

konstantintieber committed:
feat(core): Extend public API spec with retry exection endpoint

konstantintieber committed:
feat(core): Implement retry execution handler on public API

konstantintieber committed:
feat(core): Add retry execution scope to permissions

konstantintieber committed:
feat(core): Add unit tests for public retry execution endpoint

konstantintieber committed:
feat(core): Fix wrong return type in api spec

konstantintieber committed:
feat(core): Fix path to endpoint spec in openapi

konstantintieber committed:
feat(core): update test assertion

konstantintieber committed:
feat(core): Fix loadWorkflow described as parameter instead of requestBody

konstantintieber committed:
chore(core): Fix missing parameter definition in put endpoint

Found this error while adding a new endpoint to our public API

konstantintieber commented on packages/cli/src/public-api/v1/handlers/executions/executions.handler.ts:
```diff
@@ -149,4 +153,38 @@ export = {
+					publicApi: true,
+				});
+
+				return res.json({ status: retryResponse });
```
Should we return the full execution object instead? I just took the "only return new execution status" from the implementation of the internal endpoint today.

MarcL replied:
It feels like it might avoid users needing to make a 2nd call to the API for the execution data so maybe include the full response?

konstantintieber replied:
I would agree that returning the full execution object is the better API contract.
Now I see three ways to go about this:
1. Add `returnFullExecution` parameter to existing `ExecutionService.retry` to toggle returning only the new status string or the full object
2. Rename existing `ExecutionService.retry` to `retryInternal` and keep it as is. Add a new `retry` which will return the full object. We may even deprecate `retryInternal`.
3. Refactor the existing `ExecutionService.retry` to always return the full execution object. I did some digging to see who's depending on the string status return value, and it seems to me that there's only one place: [WorkflowExecutionsView::retryExection](https://github.com/n8n-io/n8n/blob/master/packages/frontend/editor-ui/src/views/WorkflowExecutionsView.vue#L286)

3 is the cleanest one imo. But will also require the most code changes.

MarcL replied:
I'm all for clean code. :) Should we do 3?

konstantintieber replied:
3 it is then.

konstantintieber commented:
Several seemingly unrelated unit tests inside `test/integration/public-api/` are now failing as a result of the changes in this PR.

From looking into [the failures](https://github.com/n8n-io/n8n/actions/runs/17457892416/job/49575908446?pr=19132), they seem to consistently return a 400 Bad Request status code now.

I don't know the reason yet.

cstuncsik committed:
fix: Add VariableId parameter reference to API specification

cstuncsik committed:
fix: Update execution retry test to use user1 and adjust response assertion

konstantintieber committed:
feat(core): update snapshot for permission scopes

konstantintieber committed:
feat(core): Make loadWorkflow docs easier to understand

konstantintieber committed:
feat(core): mock ExecutionService retry return value, since whole instance is mocked otherwise

konstantintieber committed:
Update packages/cli/src/public-api/v1/handlers/executions/spec/paths/executions.id.retry.yml

Co-authored-by: Marc Littlemore <MarcL@users.noreply.github.com>

konstantintieber committed:
feat(chore): Properly type loadWorkflow param as optional

konstantintieber committed:
feat(core): Add retry permission for execution by default for all roles

konstantintieber committed:
Add optional typing

konstantintieber committed:
refactor(core): Make retry endpoint return full execution object over just the status string (#19293)

konstantintieber committed:
Add unit tests for all known error cases on retry execution

konstantintieber committed:
Fix formatting

konstantintieber committed:
Don't include misleading data from original execution in retried execution

BGZStephen reviewed (commented):
Just a small comment around createdAt as it looked a little odd, I'm not sure about returning invalid representations for data if it isn't currently available.

konstantintieber committed:
Omit synthetic createdAt property on retried execution

We could also fetch the retried execution from the DB to get the real date.
But in our public API we also don't list createdAt as a returned field, so omitting it is fine.

konstantintieber committed:
Update API spec for retry execution endpoint

BGZStephen reviewed (approved):
The changes seem to make sense to me :+1:

Joffcom commented:
Got released with [`n8n@1.112.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.112.0)
